### DISCUSSION TITLE
CONDES in cx16-asm.cfg

### DISCUSSION THREAD

**Discussion by Ster0pes** [cc65/cc65#1407] [Q&A] [answered]
Hello! There are some questions I'd like to ask about the CONDES feature in cx16-asm.cfg:
```ini
FEATURES {
    CONDES: type    = constructor,
            label   = __CONSTRUCTOR_TABLE__,
            count   = __CONSTRUCTOR_COUNT__,
            segment = ONCE;
    CONDES: type    = destructor,
            label   = __DESTRUCTOR_TABLE__,
            count   = __DESTRUCTOR_COUNT__,
            segment = RODATA;
    CONDES: type    = interruptor,
            label   = __INTERRUPTOR_TABLE__,
            count   = __INTERRUPTOR_COUNT__,
            segment = RODATA,
            import  = __CALLIRQ__;
}
```
1) The constructor is set to segment ONCE, which is not present in this config. Shouldn't that be RODATA, like the other two entries?
2) Another question about the ONCE segment. The documentation states that segments are created when they do not exist. But, to which memory area are those segments added?
3) In order to specify an interrupt handler, a module has to export the __CALLIRQ__ symbol. Does it matter what value is exported in this symbol?
4) I tried adding a dummy constructor, destructor, and interrupt handler to an empty Assembly project. When calling cl65, I see that the symbols I marked with `.constructor`, `.destructor`, and `.interruptor` get exported, but it seems no vector tables are created. I expected to see at least the addresses of the destructor and interrupt handler in the RODATA segment. I also assumed symbols like __CONSTRUCTOR_TABLE__ would be exported, but that didn't happen either. What else do I need to do to have the linker create those vector tables?

Thanks in advance for your answers.

**Comment by oliverschmidt** [comment]
I know close to nothing about cx16 target, but nevertheless some general remarks from my side:

* Usually *-asm.cfg files don't have a `FEATURES` section at all.

* You write "The documentation states that segments are created when they do not exist." Can you please point to the exact location in the documentation saying that? From my POV the statement as-is is plainly wrong.

* The `CALLIRQ` symbol (and in fact the whole `CONDES` thingy) is from POV tied very closely to the C runtime library. I don't see any point in trying to make use of it from an assembler program.

**Reply by Ster0pes** [reply]
> You write "The documentation states that segments are created when they do not exist." Can you please point to the exact location in the documentation saying that? From my POV the statement as-is is plainly wrong.

Of course. It's in the ld65 documentation, under section '5.10 The FEATURES section', where the CONDES attribute 'segment' is explained: "This attribute tells the linker into which segment the table should be placed. **If the segment does not exist, it is created.**"

> Usually *-asm.cfg files don't have a FEATURES section at all.
> I don't see any point in trying to make use of it from an assembler program.

Well, cx16-asm.cfg does. In the ca65 documentation there's a chapter called 'Module constructors/destructors'. It starts with the following sentence:
"Note: This section applies mostly to C programs, so the explanation below uses examples from the C libraries. **However, the feature may also be useful for assembler programs.**
And indeed, I can really see its use, even in pure assembly projects. Just imagine linking several modules into your executable that expose their own set up/clean up code. The main program can iterate over all vector table entries and call all modules' constructor/destructor routines. Wouldn't that be a nice generalized way to handle a (IMHO) plausible scenario like this?

Having said all this, I'm still very new to assembly programming. I'm still at the point where I'm trying to understand the tools I just started working with. So please correct me if I'm wrong :)

**Reply by oliverschmidt** [reply]
@greg-king5: 

> 5.10 The FEATURES section', where the CONDES attribute 'segment' is explained: "This attribute tells the linker into which segment the table should be placed. If the segment does not exist, it is created."

Do you agree that the sentence "*If the segment does not exist, it is created.*" should be removed from the documentation?

**Reply by oliverschmidt** [reply]
>> Usually *-asm.cfg files don't have a FEATURES section at all.

> Well, cx16-asm.cfg does.

I know. It seems I was too ambiguous. My statement was meant to imply that I question that section in that file.

> Having said all this, I'm still very new to assembly programming. I'm still at the point where I'm trying to understand the tools I just started working with. So please correct me if I'm wrong :)

I consider that FEATURE stuff rather "advanced magic" ;-)

**Reply by greg-king5** [reply]
I removed the ambiguity:
> If the segment does not exist in any object file, it is created in the final object code.

**Reply by Ster0pes** [reply]
> I consider that FEATURE stuff rather "advanced magic" ;-)

Well, the thing is: I do understand the things being said about CONDES in the documentation. Which makes it even more frustrating/puzzling to me that I don't understand this fully :)

BTW: if you think I'm being annoying, I do apologize. The this is, I have a programming background (C#) and am really interested in trying to understand why cc65 works the way it does. I don't mean to criticize cc65 - I think it's great software.

**Comment by greg-king5** [comment, marked as answer]
> It's in the ld65 documentation, under section '5.10 The FEATURES section', where the CONDES attribute 'segment' is explained: "This attribute tells the linker into which segment the table should be placed. **If the segment does not exist, it is created.**"

I think that it means that the segment is created in the object code.  The configure file still needs to describe the segment (where it must be put, if we can put data into it in the program file, etc.).  I fixed that bug in the cx16 configure file (commit acb5af539fd5134adebb770c9ca0500195fad5ba).

In order to take advantage of the CONDES stuff (I call them "actors"), you must link your program to the "cx16.lib" library.  (The "cl65 -t cx16" command will do it automatically for you.)

Your Assembly start-up code must call the `initlib` subroutine (it runs the constructors).  Your shut-down code must call the `donelib` subroutine (it runs the destructors).

`__CALLIRQ__` is imported by ld65 _only_ if the linked code has any interruptors in it.  That name is exported by a cc65 system library module.  [That module](https://github.com/cc65/cc65/blob/master/libsrc/runtime/callirq.s) has a constructor that calls [another module](https://github.com/cc65/cc65/blob/master/libsrc/cx16/irq.s) that connects to Kernal's interrupt vector.

However, the interruptor dispatch code has some time overhead.  If you need to stuff as many actions as possible into the video chip's vertical blanking period, then you might want to write a completely custom interrupt service routine.

**Reply by Ster0pes** [reply]
First of all: if you guys think I'm  being annoying I do apologize. I do have  a programming background (no low level stuuf though, primarily C#), so I'm genuinely interested to try to understand why cc65 works the way it does. I don't mean to criticize cc65 in any way, because I think it's great software.

> In order to take advantage of the CONDES stuff (I call them "actors"), you must link your program to the "cx16.lib" library. 

I did use the -t option, so I guess I should be all set.

> Your Assembly start-up code must call the initlib subroutine (it runs the constructors). Your shut-down code must call the donelib subroutine (it runs the destructors).

In order for this to work, don't the vector tables for all constructors and destructors have to be created first, somehow?

> __CALLIRQ__ is imported by ld65 only if the linked code has any interruptors in it.

I did use the .interruptor command, doesn't that qualify? I could get cl65 to assemble my program only when I exported `__CALLIRQ__ ` myself.

**Reply by greg-king5** [reply]
> > Your Assembly start-up code must call the initlib subroutine (it runs the constructors). Your shut-down code must call the donelib subroutine (it runs the destructors).
> 
> In order for this to work, don't the vector tables for all constructors and destructors have to be created first, somehow?

ld65 creates them during linking.  They're in the program file.  A normal map file doesn't show exports that aren't imported.  You will see them after you add those calls to your program.
> 
> > `__CALLIRQ__` is imported by ld65 _only_ if the linked code has any interruptors in it.
> 
> I did use the `.interruptor` command, doesn't that qualify? I could get cl65 to assemble my program only when I exported `__CALLIRQ__` myself.

It _did_ qualify; the problem wasn't with that!  The problem was with "cx16.lib"; either it didn't export that name; or somehow, it wasn't linked.

**Reply by Ster0pes** [reply]
> A normal map file doesn't show exports that aren't imported. You will see them after you add those calls to your program.

Ah right. Importing the vector table symbols into my program does the trick indeed. I was expecting to see those symbols in the verbose map file I generated, but apparently the symbols have to be imported to show up.

As for `__CALLIRQ__`, when I import this symbol into my program, it has the value 1. Does this value have any significance, or is this just a dummy symbol? The documentation says this about the CONDES attribute import: "This can be used to force linkage of a module if this module exports the requested symbol." How does that work?

EDIT: not sure what I did wrong. Just tested it without import `__CALLIRQ__` and it still works. Still curious to know, though, how that import thing works, and why it's only needed for interruptors and not for constructors/destructors.
